### DISCUSSION TITLE
why the field is not touched when focused

### DISCUSSION THREAD

**Discussion by lauterry** [jaredpalmer/formik#2943] [General]
Hello

I wonder why the field is not marked as touched when the user **focus** on the input, even without typing anything.

Best regards

**Comment by johnrom** [comment]
Touched indicates that a field has been visited by the user. In many UI circles, receiving feedback before a user has completed their first attempt at a field is unexpected. The touched property indicates that a user may have completed their first attempt when removing focus from the field, and is often used for displaying errors. See the [ErrorMessage](https://formik.org/docs/api/errormessage) component for example.

**Comment by lauterry** [comment]
Hi

Thank for your answer. I understand.

For the behaviour illustrated below, when the user click on the input, the label goes on top, I rely on `touched` attribute to apply a specific css class for the label animation.

![input-field-label-float-to-top-on-click](https://user-images.githubusercontent.com/1374238/101062985-94a0de00-3592-11eb-9f39-b958e4e71a9f.gif)

This work as expected when the user enters something in the input but not when the user only focus the input.

My workaround is : 

```
const [helpers] = useField();

const setTouchedOnFocus =() => {
	helpers.setTouched(true);
};

<input
	onFocus={setTouchedOnFocus}
/>
```

Is there another elegant or more more straightforward way to do that ?

Best regards

**Reply by johnrom** [reply]
You can use `input:focus`.

https://developer.mozilla.org/en-US/docs/Web/CSS/:focus

**Reply by johnrom** [reply]
Alternatively you can use `onFocus` and `onBlur` _outside_ of Formik to set your class:

```jsx
const useFocus = (handleBlur, handleFocus) => {
  const [isFocused, setIsFocused] = React.useState(false);
  const onFocus = React.useCallback((event) => {
    setIsFocused(true);
    handleFocus?(event);
  }, [handleFocus]);
  const onBlur = React.useCallback((event) => {
    setIsFocused(false);
    handleBlur?(event);
  }, [handleBlur]);

  return { isFocused, onFocus, onBlur };
}

const MyField = (props) => {
  const [field] = useField(props.name);
  const {onFocus, onBlur, isFocused} = useFocus(field.handleBlur);

  return <div className={isFocused ? "focused" : ""}>
    <input {...field} onFocus={onFocus} onBlur={onBlur} />
  </div>;
}
```

**Reply by lauterry** [reply]
Hi

Thank for you reply.

This is what I end up with

**Comment by ravis-farooq** [comment]
onFocus: () => setFieldTouched('name of formik field you want to set touched property', true),

**Reply by elrouss** [reply]
A nice solution. Thank's!
